### PR TITLE
feat: Onboard EPA historical air quality dataset

### PR DESCRIPTION
**Pull request by nlarge-google**
## Description

Included: 
Annual summaries 
CO Daily Summary
CO Hourly Summary
HAP Daily Summary
HAP Hourly Summary
Lead Daily Summary
NO2 Daily Summary
NO2 Hourly Summary
NONOxNOy Daily Summary
NONOxNOy Hourly Summary
Ozone Daily Summary
Ozone Hourly Summary
PM 10 Daily Summary
PM10 Hourly Summary
PM25 Frm Hourly Summary
PM25 NonFrm Daily Summary
PM25 NonFrm Hourly Summary
PM25 Speciation Daily Summary
PM25 Speciation Hourly Summary
Pressure Daily Summary
Pressure Hourly Summary
RH and DP Daily Summary
RH and DP Hourly Summary
SO2 Daily Summary
SO2 Hourly Summary
Temperature Daily Summary
Temperature Hourly Summary
VOC Daily Summary
VOC Hourly Summary
Wind Daily Summary
Wind Hourly Summary

## Checklist

Note: Delete items below that aren't applicable to your pull request.

- [x] Please merge this PR for me once it is approved.
- [x] If this PR adds or edits a dataset or pipeline, it was reviewed and approved by the Google Cloud Public Datasets team beforehand.
- [x] If this PR adds or edits a dataset or pipeline, I put all my code inside  `datasets/epa_historical_air_quality` and nothing outside of that directory.
- [x] This PR is appropriately labeled.

### CODE REVIEW

**Comment by nlarge-google** [comment]
@adlersantos Ready for review.  Thanks!

File: datasets/epa_historical_air_quality/_images/run_csv_transform_kub/csv_transform.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+    for col in df.columns:
+        if df[col].dtype == "datetime64[ns]":
+            logging.info(f"Resolving datetime on {col}")
+            df[col] = df[col].apply(lambda x: convert_dt_format(str(x), from_format))
+
+    return df
+
+
+def convert_dt_format(dt_str: str, from_format: str) -> str:
+    # rtnval = ""
```
Comment (adlersantos) [inline datasets/epa_historical_air_quality/_images/run_csv_transform_kub/csv_transform.py:219]: Extra line?

File: datasets/epa_historical_air_quality/_images/run_csv_transform_kub/csv_transform.py [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+        len(from_format.strip().split("/")[0]) == 2
+    ):
+        # if the format of the date portion of the data is in YYYY-MM-DD format
+        # and from_format is in MM-DD-YYYY then resolve this by modifying the from_format
+        # to use the YYYY-MM-DD.  This resolves mixed date formats in files
+        from_format = "%Y-%m-%d " + from_format.strip().split(" ")[1]
+    else:
+        dt_str = ""
+
+    # return datetime.datetime.strptime(dt_str, from_format).strftime("%Y-%m-%d %H:%M:%S")
```
Comment (adlersantos) [inline datasets/epa_historical_air_quality/_images/run_csv_transform_kub/csv_transform.py:241]: Extra line

File: datasets/epa_historical_air_quality/_images/run_csv_transform_kub/csv_transform.py [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+
+def save_to_new_file(df, file_path, sep="|") -> None:
+    logging.info(f"Saving to file {file_path} separator='{sep}'")
+    df.to_csv(file_path, sep=sep, index=False)
+
+
+def append_batch_file(
+    batch_file_path: str, target_file_path: str, skip_header: bool, truncate_file: bool
+) -> None:
+    data_file = open(batch_file_path, "r")
```
Comment (adlersantos) [inline datasets/epa_historical_air_quality/_images/run_csv_transform_kub/csv_transform.py:253]: Another way to do this is using context blocks:
```
with open(batch_file_path, "r") as data_file:
   # code here
```
This prevents from having to manually close the file via `data_file.close()`.

File: datasets/epa_historical_air_quality/_images/run_csv_transform_kub/csv_transform.py [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+    if truncate_file:
+        target_file = open(target_file_path, "w+").close()
+    target_file = open(target_file_path, "a+")
```
Comment (adlersantos) [inline datasets/epa_historical_air_quality/_images/run_csv_transform_kub/csv_transform.py:256]: If you have to files that need to be open, you can nest the context blocks

File: datasets/epa_historical_air_quality/_images/run_csv_transform_kub/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,5 @@
+requests
+numpy
+pandas
+google-cloud-storage
+gsutil
```
Comment (adlersantos) [inline datasets/epa_historical_air_quality/_images/run_csv_transform_kub/requirements.txt:5]:
```suggestion
pandas
google-cloud-storage
gsutil
```

File: datasets/epa_historical_air_quality/annual_summaries/pipeline.yaml [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+              nodeSelectorTerms:
+                - matchExpressions:
+                    - key: cloud.google.com/gke-nodepool
+                      operator: In
+                      values:
+                        - "pool-e2-standard-4"
+        image_pull_policy: "Always"
+        image: "{{ var.json.epa_historical_air_quality.container_registry.run_csv_transform_kub }}"
+        env_vars:
+          SOURCE_URL: "https://aqs.epa.gov/aqsweb/airdata/annual_conc_by_monitor_~year~.zip"
```
Comment (adlersantos) [inline datasets/epa_historical_air_quality/annual_summaries/pipeline.yaml:57]:
```suggestion
          SOURCE_URL: "https://aqs.epa.gov/aqsweb/airdata/annual_conc_by_monitor_YEAR.zip"
```

File: datasets/epa_historical_air_quality/co_daily_summary/pipeline.yaml [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+              nodeSelectorTerms:
+                - matchExpressions:
+                    - key: cloud.google.com/gke-nodepool
+                      operator: In
+                      values:
+                        - "pool-e2-standard-4"
+        image_pull_policy: "Always"
+        image: "{{ var.json.epa_historical_air_quality.container_registry.run_csv_transform_kub }}"
+        env_vars:
+          SOURCE_URL: "https://aqs.epa.gov/aqsweb/airdata/daily_42101_~year~.zip"
```
Comment (adlersantos) [inline datasets/epa_historical_air_quality/co_daily_summary/pipeline.yaml:57]: ^

File: datasets/epa_historical_air_quality/co_hourly_summary/pipeline.yaml [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+              nodeSelectorTerms:
+                - matchExpressions:
+                    - key: cloud.google.com/gke-nodepool
+                      operator: In
+                      values:
+                        - "pool-e2-standard-4"
+        image_pull_policy: "Always"
+        image: "{{ var.json.epa_historical_air_quality.container_registry.run_csv_transform_kub }}"
+        env_vars:
+          SOURCE_URL: "https://aqs.epa.gov/aqsweb/airdata/hourly_42101_~year~.zip"
```
Comment (adlersantos) [inline datasets/epa_historical_air_quality/co_hourly_summary/pipeline.yaml:57]: ^ or use `YEAR_ITERATION` to show the intention that a part of the URL is going to be iterated upon

**Review by adlersantos — CHANGES_REQUESTED** [review]
Looks good in general. Just a few comments.

**Comment by nlarge-google** [comment]
@adlersantos I have done research into trying to understand why this PR keeps failing at the Run Tests point.  Is there some recent change to the code behind these checks?  What changes do I need to make to ensure that these checks pass? Thanks!

**Comment by nlarge-google** [comment]
@adlersantos I dont see any button in github UI that allows me to rerun the checks without making code changes, so I had to close and reopen the PR in order to trigger the tests.  If you know a better way to re-trigger the checks, or if closing and re-opening the PR is acceptable then please let me know.

File: datasets/epa_historical_air_quality/_images/run_csv_transform_kub/csv_transform.py (hunk 1 above)
Comment (nlarge-google) [reply datasets/epa_historical_air_quality/_images/run_csv_transform_kub/csv_transform.py]: Done.  Removed the line.

File: datasets/epa_historical_air_quality/_images/run_csv_transform_kub/csv_transform.py (hunk 2 above)
Comment (nlarge-google) [reply datasets/epa_historical_air_quality/_images/run_csv_transform_kub/csv_transform.py:240]: Done

**Comment by nlarge-google** [comment]
@adlersantos Please review this code at your earliest convenience.  Thanks!

File: datasets/epa_historical_air_quality/_images/run_csv_transform_kub/csv_transform.py (hunk 3 above)
Comment (nlarge-google) [reply datasets/epa_historical_air_quality/_images/run_csv_transform_kub/csv_transform.py]: Done.

File: datasets/epa_historical_air_quality/_images/run_csv_transform_kub/csv_transform.py (hunk 4 above)
Comment (nlarge-google) [reply datasets/epa_historical_air_quality/_images/run_csv_transform_kub/csv_transform.py]: Done

File: datasets/epa_historical_air_quality/_images/run_csv_transform_kub/requirements.txt (hunk 5 above)
Comment (nlarge-google) [reply datasets/epa_historical_air_quality/_images/run_csv_transform_kub/requirements.txt]: Done

File: datasets/epa_historical_air_quality/annual_summaries/pipeline.yaml (hunk 6 above)
Comment (nlarge-google) [reply datasets/epa_historical_air_quality/annual_summaries/pipeline.yaml]: Done, although I put YEAR_ITERATOR, not YEAR.  Hope that works.  If not I can redo.

**Comment by nlarge-google** [comment]
@adlersantos I have made the changes.  Please review.  Thanks.

**Comment by nlarge-google** [comment]
@adlersantos All requested changes are completed.  Please review again.  Thanks!

**Review by adlersantos — APPROVED** [review]
Thanks a lot!

**Pull request merged**